### ISSUE TITLE
DirectoryEntry.GetName assert failure on macOS

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#27038]
https://mc.dot.net/#/user/stephentoub/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/a137bbf42598202f999ead31d92754613145c849/workItem/System.IO.Tests/wilogs
```
2018-02-10 19:37:45,431: INFO: proc(54): run_and_log_output: Output: Discovering: System.IO.Tests
2018-02-10 19:37:45,666: INFO: proc(54): run_and_log_output: Output: Discovered:  System.IO.Tests
2018-02-10 19:37:45,821: INFO: proc(54): run_and_log_output: Output: Starting:    System.IO.Tests
2018-02-10 19:37:46,351: INFO: proc(54): run_and_log_output: Output: Assertion Failed
2018-02-10 19:37:46,351: INFO: proc(54): run_and_log_output: Output: should not have embedded nulls
2018-02-10 19:37:46,351: INFO: proc(54): run_and_log_output: Output: 
2018-02-10 19:37:46,351: INFO: proc(54): run_and_log_output: Output:    at Interop.Sys.DirectoryEntry.GetName(Span`1 buffer)
2018-02-10 19:37:46,351: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEntry.get_FileName()
2018-02-10 19:37:46,352: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEnumerableFactory.<>c__DisplayClass5_0.<UserEntries>b__1(FileSystemEntry& entry)
2018-02-10 19:37:46,352: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEnumerable`1.DelegateEnumerator.ShouldIncludeEntry(FileSystemEntry& entry)
2018-02-10 19:37:46,352: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()
2018-02-10 19:37:46,352: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileSystem.RemoveDirectoryInternal(DirectoryInfo directory, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
2018-02-10 19:37:46,353: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
2018-02-10 19:37:46,353: INFO: proc(54): run_and_log_output: Output:    at System.IO.Directory.Delete(String path, Boolean recursive)
2018-02-10 19:37:46,353: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileCleanupTestBase.Dispose(Boolean disposing) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/CoreFx.Private.TestUtilities/src/System/IO/FileCleanupTestBase.cs:line 66
2018-02-10 19:37:46,354: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileCleanupTestBase.Dispose() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/CoreFx.Private.TestUtilities/src/System/IO/FileCleanupTestBase.cs:line 57
2018-02-10 19:37:46,354: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExecutionTimer.Aggregate(Action action)
2018-02-10 19:37:46,354: INFO: proc(54): run_and_log_output: Output:    at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource)
2018-02-10 19:37:46,354: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.Run(Action code)
2018-02-10 19:37:46,354: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestInvoker`1.<RunAsync>b__45_0()
2018-02-10 19:37:46,355: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,355: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestInvoker`1.<RunAsync>b__45_0()
2018-02-10 19:37:46,355: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,355: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,356: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,356: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitTestRunner.InvokeTestAsync(ExceptionAggregator aggregator)
2018-02-10 19:37:46,356: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,356: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitTestRunner.InvokeTestAsync(ExceptionAggregator aggregator)
2018-02-10 19:37:46,356: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,357: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,357: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,357: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestRunner`1.RunAsync()
2018-02-10 19:37:46,357: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,358: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestRunner`1.RunAsync()
2018-02-10 19:37:46,358: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitTheoryTestCaseRunner.RunTestAsync()
2018-02-10 19:37:46,358: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,358: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitTheoryTestCaseRunner.RunTestAsync()
2018-02-10 19:37:46,359: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCaseRunner`1.RunAsync()
2018-02-10 19:37:46,359: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,359: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCaseRunner`1.RunAsync()
2018-02-10 19:37:46,359: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunTestCasesAsync()
2018-02-10 19:37:46,360: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,360: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunTestCasesAsync()
2018-02-10 19:37:46,360: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunAsync()
2018-02-10 19:37:46,360: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,361: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunAsync()
2018-02-10 19:37:46,361: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunTestMethodsAsync()
2018-02-10 19:37:46,361: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,361: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunTestMethodsAsync()
2018-02-10 19:37:46,361: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunAsync()
2018-02-10 19:37:46,362: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,362: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunAsync()
2018-02-10 19:37:46,362: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunTestClassesAsync()
2018-02-10 19:37:46,362: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,362: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunTestClassesAsync()
2018-02-10 19:37:46,363: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunAsync()
2018-02-10 19:37:46,363: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,363: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunAsync()
2018-02-10 19:37:46,363: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task`1.InnerInvoke() in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/future.cs:line 610
2018-02-10 19:37:46,364: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:46,364: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/Task.cs:line 2440
2018-02-10 19:37:46,364: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteEntry() in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/Task.cs:line 2361
2018-02-10 19:37:46,365: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.SynchronizationContextTaskScheduler.<>c.<.cctor>b__8_0(Object s) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/TaskScheduler.cs:line 694
2018-02-10 19:37:46,365: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.MaxConcurrencySyncContext.RunOnSyncContext(SendOrPostCallback callback, Object state)
2018-02-10 19:37:46,365: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:46,365: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.MaxConcurrencySyncContext.WorkerThreadProc()
2018-02-10 19:37:46,365: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitWorkerThread.<>c.<QueueUserWorkItem>b__5_0(Object _)
2018-02-10 19:37:46,366: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:46,366: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/Task.cs:line 2440
2018-02-10 19:37:46,366: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:46,367: INFO: proc(54): run_and_log_output: Output: Assertion Failed
2018-02-10 19:37:46,367: INFO: proc(54): run_and_log_output: Output: should not have embedded nulls
2018-02-10 19:37:46,367: INFO: proc(54): run_and_log_output: Output: 
2018-02-10 19:37:46,367: INFO: proc(54): run_and_log_output: Output:    at Interop.Sys.DirectoryEntry.GetName(Span`1 buffer)
2018-02-10 19:37:46,367: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEntry.get_FileName()
2018-02-10 19:37:46,367: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEnumerableFactory.<>c__DisplayClass5_0.<UserEntries>b__1(FileSystemEntry& entry)
2018-02-10 19:37:46,367: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEnumerable`1.DelegateEnumerator.ShouldIncludeEntry(FileSystemEntry& entry)
2018-02-10 19:37:46,368: INFO: proc(54): run_and_log_output: Output:    at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()
2018-02-10 19:37:46,368: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileSystem.RemoveDirectoryInternal(DirectoryInfo directory, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
2018-02-10 19:37:46,368: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
2018-02-10 19:37:46,368: INFO: proc(54): run_and_log_output: Output:    at System.IO.Directory.Delete(String path, Boolean recursive)
2018-02-10 19:37:46,368: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileCleanupTestBase.Dispose(Boolean disposing) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/CoreFx.Private.TestUtilities/src/System/IO/FileCleanupTestBase.cs:line 66
2018-02-10 19:37:46,369: INFO: proc(54): run_and_log_output: Output:    at System.IO.FileCleanupTestBase.Dispose() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/CoreFx.Private.TestUtilities/src/System/IO/FileCleanupTestBase.cs:line 57
2018-02-10 19:37:46,369: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExecutionTimer.Aggregate(Action action)
2018-02-10 19:37:46,369: INFO: proc(54): run_and_log_output: Output:    at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource)
2018-02-10 19:37:46,369: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.Run(Action code)
2018-02-10 19:37:46,369: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestInvoker`1.<RunAsync>b__45_0()
2018-02-10 19:37:46,370: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,370: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestInvoker`1.<RunAsync>b__45_0()
2018-02-10 19:37:46,370: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,370: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,371: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,371: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitTestRunner.InvokeTestAsync(ExceptionAggregator aggregator)
2018-02-10 19:37:46,371: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,371: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitTestRunner.InvokeTestAsync(ExceptionAggregator aggregator)
2018-02-10 19:37:46,371: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,372: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,372: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code)
2018-02-10 19:37:46,372: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestRunner`1.RunAsync()
2018-02-10 19:37:46,372: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,372: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestRunner`1.RunAsync()
2018-02-10 19:37:46,372: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCaseRunner`1.RunAsync()
2018-02-10 19:37:46,373: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,373: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCaseRunner`1.RunAsync()
2018-02-10 19:37:46,373: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunTestCasesAsync()
2018-02-10 19:37:46,373: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,374: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunTestCasesAsync()
2018-02-10 19:37:46,374: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunAsync()
2018-02-10 19:37:46,374: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,374: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestMethodRunner`1.RunAsync()
2018-02-10 19:37:46,374: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunTestMethodsAsync()
2018-02-10 19:37:46,375: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,375: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunTestMethodsAsync()
2018-02-10 19:37:46,375: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunAsync()
2018-02-10 19:37:46,375: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,375: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestClassRunner`1.RunAsync()
2018-02-10 19:37:46,375: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunTestClassesAsync()
2018-02-10 19:37:46,376: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,376: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunTestClassesAsync()
2018-02-10 19:37:46,376: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunAsync()
2018-02-10 19:37:46,376: INFO: proc(54): run_and_log_output: Output:    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 922
2018-02-10 19:37:46,376: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.TestCollectionRunner`1.RunAsync()
2018-02-10 19:37:46,377: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task`1.InnerInvoke() in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/future.cs:line 610
2018-02-10 19:37:46,377: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:46,377: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/Task.cs:line 2440
2018-02-10 19:37:46,377: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteEntry() in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/Task.cs:line 2361
2018-02-10 19:37:46,378: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.SynchronizationContextTaskScheduler.<>c.<.cctor>b__8_0(Object s) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/TaskScheduler.cs:line 694
2018-02-10 19:37:46,378: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.MaxConcurrencySyncContext.RunOnSyncContext(SendOrPostCallback callback, Object state)
2018-02-10 19:37:46,378: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:46,378: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.MaxConcurrencySyncContext.WorkerThreadProc()
2018-02-10 19:37:46,379: INFO: proc(54): run_and_log_output: Output:    at Xunit.Sdk.XunitWorkerThread.<>c.<QueueUserWorkItem>b__5_0(Object _)
2018-02-10 19:37:46,379: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:46,379: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot) in /Users/buildagent/agent/_work/382/s/src/mscorlib/src/System/Threading/Tasks/Task.cs:line 2440
2018-02-10 19:37:46,379: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /Users/buildagent/agent/_work/382/s/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 166
2018-02-10 19:37:49,001: INFO: proc(54): run_and_log_output: Output: /Users/dotnet-bot/dotnetbuild/work/78908c4b-3dbb-4ac4-8785-16df253cd217/Work/5674dd81-888b-46a4-9774-b7b0136812e8/Unzip/RunTests.sh: line 87: 64175 Abort trap: 6   
```
cc: @JeremyKuhne

**Comment by danmosemsft** [comment]
I just hit this on Centos. running some tests in a loop.
```
Assertion Failed
should not have embedded nulls if we parsed the end of string

   at Interop.Sys.DirectoryEntry.GetName(Span`1 buffer)
   at System.IO.Enumeration.FileSystemEntry.get_FileName()
   at System.IO.Enumeration.FileSystemEntry.get_FullPath()
   at System.IO.Enumeration.FileSystemEntry.Initialize(FileSystemEntry& entry, DirectoryEntry directoryEntry, ReadOnlySpan`1 directory, String rootDirectory, String originalRootDirectory, Span`1 pathBuffer)
   at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()
   at System.IO.FileSystem.RemoveDirectoryInternal(DirectoryInfo directory, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
   at System.IO.Directory.Delete(String path, Boolean recursive)
   at System.IO.FileCleanupTestBase.Dispose(Boolean disposing) in /mnt/resource/workspace/rhel7.2_release/src/CoreFx.Private.TestUtilities/src/System/IO/FileCleanupTestBase.cs:line 66
   at System.IO.FileCleanupTestBase.Dispose() in /mnt/resource/workspace/rhel7.2_release/src/CoreFx.Private.TestUtilities/src/System/IO/FileCleanupTestBase.cs:line 57
   at Xunit.Sdk.ExecutionTimer.Aggregate(Action action)
   at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource)

```

**Comment by JeremyKuhne** [comment]
@danmosemsft what is your current head?

**Comment by danmosemsft** [comment]
This is a repro box. Seems to be 46267c1dbe46f0127e670ce25933ff49a5ec9d6e

**Comment by JeremyKuhne** [comment]
Hmm. Again in FileCleanupTestBase... Possibly a handle reuse? I can't find a way we would get into that state, however. I'll add some more details to the assert.

@tarekgh is it possible for `Encoding.UTF8.GetChars(nameBytes, buffer);`  to return an embedded null char if you pass it an array of bytes with no zeroes?

**Comment by JeremyKuhne** [comment]
Link to the asserting code: https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Unix/System.Native/Interop.ReadDir.cs#L37-L56

**Comment by danmosemsft** [comment]
I can repro easily. Tell me what you want me to put in the assert

**Comment by tarekgh** [comment]
> is it possible for Encoding.UTF8.GetChars(nameBytes, buffer); to return an embedded null char if you pass it an array of bytes with no zeroes?

I don't think so.

**Comment by danmosemsft** [comment]
@JeremyKuhne if you push changes/asserts up to some branch, I can test them for you. It hits in a bout 10 minutes.

**Comment by danmosemsft** [comment]
On the Centos machine we found
* If we dump value.ToString() if the assert fires, we consistently segfault.
* In one case we got a hit higher up at `                Debug.Assert(nameBytes.Length > 0, "we shouldn't have gotten a garbage value from the OS");`
* On the `                Debug.Assert(NameLength != -1 || value.IndexOf('\0') == -1, "should not have embedded nulls if we parsed the end of string");` line we dumped charCount, value.Length and bytes.Length and we got a hit where they were all the same (58). We also dumped value.IndexOf('\0') and it was 0.

Next steps
* Try to repro with chk clrjit.dll to see whether there are interesting asserts
* Put the ToString() back so it segfaults, and catch in a debugger (or get a dump)
